### ISSUE TITLE
Add support for immutable tensors in torch.export

### ISSUE THREAD

**Issue by dgcnz** [pytorch/pytorch#136642]
### 🚀 The feature, motivation and pitch

I'm working on exporting an object detection model which casts a list of tensor shapes into a tensor, and I'm trying to hint the compiler that it should treat it as a constant in order to avoid unnecessary `GuardOnDataDependentSymNode` errors. The problem is that when the list is casted to a tensor, it's data "unspecializes" and throws `GuardOnDataDependentSymNode` errors (on the linspace line). 

```py
import torch
import torch.nn as nn
import typing
import torch._dynamo as dynamo
import logging

class NetMRE(nn.Module):
    def forward(self, x: typing.List[torch.Tensor]):
        spatial_shapes: typing.List[typing.Tuple[int, int]] = []
        for xi in x:
            spatial_shapes.append(xi.shape[2:])

        spatial_shapes = torch.as_tensor(spatial_shapes, dtype=torch.long)
        reference_points_list = []
        for H, W in spatial_shapes:
            lin = torch.linspace(0.5, H - 0.5, H, dtype=torch.float32) # --- breaks here
            reference_points_list.append(lin)
        return reference_points_list

example_kwargs = {
    "x": [
        torch.rand(1, 3, 64, 64),
        torch.rand(1, 3, 32, 32),
    ],
}
exported_program: torch.export.ExportedProgram = export(
    NetMRE(), (), kwargs=example_kwargs, strict=True,
)```

### Alternatives

1. The easiest fix is to never cast the list to tensor and refactor all subsequent code that uses the tensor operations on that object. Although this is not a terrible alternative, I feel like it could be easily avoidable with the suggested feature.

2. The hackier solution I've found so far is disable the faketensor creation, but it doesn't work with `strict=True` and it also seems like it would probably break other things.

```py
class NetC2v2(nn.Module):
    def forward(self, x: typing.List[torch.Tensor]):
        spatial_shapes: typing.List[typing.Tuple[int, int]] = []
        for i, xi in enumerate(x):
            spatial_shapes.append(tuple(xi.shape[2:]))
        
        with unset_fake_temporarily():
            with disable_proxy_modes_tracing():
                spatial_shapes_tensor = torch.tensor(spatial_shapes, dtype=torch.long)
        reference_points = self.get_reference_points(spatial_shapes_tensor)
        return reference_points

    @staticmethod
    def get_reference_points(
        spatial_shapes: torch.Tensor,
    ): 
        reference_points_list = []
        for i in range(len(spatial_shapes)):
            with disable_proxy_modes_tracing():
                with unset_fake_temporarily():
                    h = int(spatial_shapes[i][0])
            lin = torch.linspace(0.5, h - 0.5, h, dtype=torch.float32)
            reference_points_list.append(lin)
        return reference_points_list
```

**Comment by ezyang** [comment]
Yeah, fake tensor has constant prop, we just have to make sure it actually gets triggered here. Your tensor is small enough that it should be eligible, so long as the `as_tensor` call appropriately passes make_constant=True to from_real_tensor. Have you tried using torch.tensor instead? I think maybe that one is setup.

**Comment by dgcnz** [comment]
Yes, previously I had seen that constant scalar tensors (`t.numel() <= 1`) were supported:

https://github.com/pytorch/pytorch/blob/9656a603b24ba1e3cdeae8839ef18f6364c458f5/torch/_subclasses/fake_tensor.py#L1756-L1764

https://github.com/pytorch/pytorch/blob/9656a603b24ba1e3cdeae8839ef18f6364c458f5/torch/_subclasses/fake_tensor.py#L2220-L2226

https://github.com/pytorch/pytorch/blob/9656a603b24ba1e3cdeae8839ef18f6364c458f5/torch/_subclasses/fake_tensor.py#L101

Although, for some reason it never occurred to me to override that constant...

So, in conclusion: `torch._subclasses.fake_tensor.CONSTANT_NUMEL_LIMIT = 8` before compilation works :^)

Thanks @ezyang ! 

PS: Do you know if this could have any unwanted side effects? Why is `CONSTANT_NUMEL_LIMIT` set to such a small value? (1)

**Comment by eellison** [comment]
We could consider moving it higher. It's low pessimistically because of memory use concerns. Currently, I don't think we ever clear the constants from run to run so recompilations would accrue memory. Although we could clear all the fake tensors constacks in the stack after inductor is done.

Not sure what is a good default.

**Comment by eellison** [comment]
This is also something the compute_at_compile_time could help with, I forget exactly what the api is

**Comment by ezyang** [comment]
Let's bump it higher / make it user configurable. 

The option is have a way to explicitly say "please keep constant propagating this no matter how large until we're done"

**Comment by dgcnz** [comment]
@ezyang I noted that not all operations on this constant tensor succeed. For example, calling `.item()` makes it fail with `Unsupported: torch.* op returned non-Tensor int call_method item`:

New script (only change is `.item()`, full logs in this [gist](https://gist.github.com/dgcnz/cbbd25f8e2e9a4c0b0fc4f239b19c323)):
```py
import torch
import torch.nn as nn
import typing
import torch._dynamo as dynamo
import logging

class NetMRE(nn.Module):
    def forward(self, x: typing.List[torch.Tensor]):
        spatial_shapes: typing.List[typing.Tuple[int, int]] = []
        for xi in x:
            spatial_shapes.append(xi.shape[2:])

        spatial_shapes = torch.as_tensor(spatial_shapes, dtype=torch.long)
        reference_points_list = []
        for H, W in spatial_shapes:
            lin = torch.linspace(0.5, H.item() - 0.5, H.item(), dtype=torch.float32) # --- Adding .item() breaks
            reference_points_list.append(lin)
        return reference_points_list

example_kwargs = {
    "x": [
        torch.rand(1, 3, 64, 64),
        torch.rand(1, 3, 32, 32),
    ],
}
torch._subclasses.fake_tensor.CONSTANT_NUMEL_LIMIT = 8
exported_program: torch.export.ExportedProgram = torch.export.export(
    NetMRE(), (), kwargs=example_kwargs, strict=True,
)
```


The relevant torch snippet is this: 
https://github.com/pytorch/pytorch/blob/2980aed65b6c521e41ec8a995f4c94f184dd741b/torch/_dynamo/variables/builder.py#L2390-L2395

Adding the following line just before the previous snippet works, but there's probably a more graceful way haha:

```py
elif proxy.node.target == 'item' and proxy.node.args[0].meta["example_value"].constant:
        return ConstantVariable.create(example_value, **options)
```

**Comment by ezyang** [comment]
That's an independent Dynamo bug, where you have to explicitly say that it's OK for functions to return integer constants. I've tried universally accepting integer outputs but this is apparently load bearing so some fixing needed. WAR is fine...

**Comment by dgcnz** [comment]
@ezyang Okay, should I report these bugs in separate issues? 

Also, these tensors throw "Dynamic slicing on data-dependent value is not supported" when one attempts to index another tensor with them (minimum reproducibility code in this [gist](https://gist.github.com/dgcnz/77d39c8a8177bd827c71d9aa6f191fcf)).

The relevant torch snippet is this:

https://github.com/pytorch/pytorch/blob/2265c2d48cc7425a9cb3d64f4d469e71979ce136/torch/_dynamo/variables/lists.py#L811-L814

Which similarly can be amended by replacing it with something like this:

```python
        if isinstance(start, variables.TensorVariable):
            try:
                start = ConstantVariable.create(start.proxy.node.meta["example_value"].constant.item())
            except Exception:
                unimplemented("Dynamic slicing on data-dependent value is not supported")
        if isinstance(stop, variables.TensorVariable):
            try:
                stop = ConstantVariable.create(stop.proxy.node.meta["example_value"].constant.item())
            except Exception:
                unimplemented("Dynamic slicing on data-dependent value is not supported")
        if isinstance(step, variables.TensorVariable):
            try:
                step = ConstantVariable.create(step.proxy.node.meta["example_value"].constant.item())
            except Exception:
                unimplemented("Dynamic slicing on data-dependent value is not supported")
```

**Comment by ezyang** [comment]
Feel free to do them separately if they have different texture / error message. However, your latest example also will be fixed by more aggressive const propping, no need to edit each site individually.

**Comment by laithsakka** [comment]
@dgcnz  dynamic slicing shall be supported now
